src/wast-parser.cc

```diff
@@ -0,0 +1,2013 @@
+    if (Failed(ParseUint64(sv.begin(), sv.end(), &index))) {
+      // Print an error, but don't fail parsing.
+      Error(token.loc, "invalid int \"" PRIstringview "\"",
+            WABT_PRINTF_STRING_VIEW_ARG(sv));
+    }
+
+    *out_var = Var(index, token.loc);
+    return Result::Ok;
+  } else if (PeekMatch(TokenType::Var)) {
+    auto token = Consume();
+    *out_var = Var(token.text.to_string_view(), token.loc);
+    return Result::Ok;
+  } else {
+    return ErrorExpected({"a numeric index", "a name"}, "12 or $foo");
+  }
+}
+
+bool WastParser::ParseVarOpt(Var* out_var, Var default_var) {
+  WABT_TRACE(ParseVarOpt);
+  if (PeekMatch(TokenType::Nat) || PeekMatch(TokenType::Var)) {
+    Result result = ParseVar(out_var);
+    // Should always succeed, the only way it could fail is if the token
+    // doesn't match.
+    assert(Succeeded(result));
+    WABT_USE(result);
+    return true;
+  } else {
+    *out_var = default_var;
+    return false;
+  }
+}
+
+Result WastParser::ParseOffsetExpr(ExprList* out_expr_list) {
+  WABT_TRACE(ParseOffsetExpr);
+  if (MatchLpar(TokenType::Offset)) {
+    CHECK_RESULT(ParseTerminatingInstrList(out_expr_list));
+    EXPECT(Rpar);
+  } else if (PeekMatchExpr()) {
+    CHECK_RESULT(ParseExpr(out_expr_list));
+  } else {
+    return ErrorExpected({"an offset expr"}, "(i32.const 123)");
+  }
+  return Result::Ok;
+}
+
+Result WastParser::ParseTextList(std::vector<uint8_t>* out_data) {
+  WABT_TRACE(ParseTextList);
+  if (!ParseTextListOpt(out_data)) {
+    return Result::Error;
+  }
+
+  return Result::Ok;
+}
+
+bool WastParser::ParseTextListOpt(std::vector<uint8_t>* out_data) {
+  WABT_TRACE(ParseTextListOpt);
+  TextVector texts;
+  while (PeekMatch(TokenType::Text))
+    texts.push_back(Consume().text.to_string());
+
+  RemoveEscapes(texts, std::back_inserter(*out_data));
+  return !texts.empty();
+}
+
+Result WastParser::ParseVarList(VarVector* out_var_list) {
+  WABT_TRACE(ParseVarList);
+  if (!ParseVarListOpt(out_var_list)) {
+    return Result::Error;
+  }
+
+  return Result::Ok;
+}
+
+bool WastParser::ParseVarListOpt(VarVector* out_var_list) {
+  WABT_TRACE(ParseVarListOpt);
+  Var var;
+  while (ParseVarOpt(&var))
+    out_var_list->push_back(var);
+
+  return !out_var_list->empty();
+}
+
+Result WastParser::ParseValueType(Type* out_type) {
+  WABT_TRACE(ParseValueType);
+  if (!PeekMatch(TokenType::ValueType))
+    return ErrorExpected({"i32", "i64", "f32", "f64"});
+
+  *out_type = Consume().type;
+  return Result::Ok;
+}
+
+Result WastParser::ParseValueTypeList(TypeVector* out_type_list) {
+  WABT_TRACE(ParseValueTypeList);
+  while (PeekMatch(TokenType::ValueType))
+    out_type_list->push_back(Consume().type);
+
+  CHECK_RESULT(ErrorIfLpar({"i32", "i64", "f32", "f64"}));
+  return Result::Ok;
+}
+
+Result WastParser::ParseQuotedText(std::string* text) {
+  WABT_TRACE(ParseQuotedText);
+  if (!PeekMatch(TokenType::Text))
+    return ErrorExpected({"a quoted string"}, "\"foo\"");
```

Comment:
This is just for example text. Maybe not needed, but I thought it made the errors a little nicer:
```
error: unexpected token "(", expected a quoted string (e.g. "foo").
```